### PR TITLE
Move the essentials of IO.chpl into ChapelIO and make the rest not included by default

### LINKED ISSUE #13448: proposal: do not include IO module by default but enable writeln
This issue started over on  #13373.

This issue proposes that we move the top-level free functions `writeln` and `write` to `ChapelIO.chpl` (which is an internal module that has its contents visible by default) and then that we make everything else that is in `IO.chpl` not be visible by default. Please note that these two functions do not throw (instead they halt by default). The channel type, `stderr`, stdout, methods on channels, file opening, etc. would not be included by default.

The idea here would be:
 1. Have programs indicate, with `use IO;` or similar, when they are making use of non-trivial I/O features
 2. Reduce namespace pollution for programs that only use writeln from IO.chpl
 3. Enable IO to make visible the errors in SysError that are thrown by it but without making SysError errors visible everywhere by default.

### PR DESCRIPTION
**Pull request by lydia-duncan**
As proposed in #13448, this PR will limit the IO related symbols that are
visible by default to merely `write`, `writeln`, and `writef`, and make the use
of IO in ChapelIO `private`.

Updated the documentation for this change, including updating links to the
symbols and removing the note at the top of `IO` saying that it was
included by default.  While here, updated the documentation for `write`,
`read`, `writeln`, and `readln` to use code highlighting around the
`try! stdout.*` section.

Updates a whole bunch of tests and modules for this new behavior.

Of note: while the spec Error handling tests for catching errors rely on a
function defined in another .chpl file, the error they "catch" was not actually
used in that function, so I added a use of SysError to the client files instead
of the file defining the function.

Also of note: now that writeln is defined in an internal module instead of a
standard one, it is no longer printed by the compiler flag `--print-callgraph`.

Also of note: many modules with explicit read/writeThis methods now have
a private use of the IO module.  This is because, while those methods are
currently documented and take a `channel` argument, they are only intended
to be used under the covers rather than explicitly called by users.  It would be
good to remove the documentation and "no doc" them to reflect this, but in most
cases I did not undertake that effort as part of this PR.

Resolves #13448 

Deprecation warnings:
I've added a module LastResortIO.  This module contains functions with the same
name as symbols defined by the IO module, which generate a deprecation warning
indicating that the symbols are no longer available by default and then performs
the original action.  In some cases, it is not possible to enable the original
action after the message, and so a deprecation error is used instead.

The symbols it covers are:
    - openmem
    - openreader
    - opentmp
    - openfp
    - openfd
    - open
    - iohints
    - defaultIOStyle
    - IOHINT_*
    - stringStyleWithVariableLength
    - stringStyleWithLength
    - stringStyleExactLen
    - stringStyleNullTerminated
    - iolittle
    - iobig
    - ionative
    - iodynamic
    - iokind
    - iostringformat
    - iostringstyle
    - iomode
and briefly file and some of its methods, as well as a channel method

This file is a work in progress - I have TODOs for deprecation warnings I have
not tried to implement yet, and have commented out the deprecation warnings for
symbols I have not yet tested, due to concerns about generating unhelpful
messages.  I intend to continue working on this but don't think it is reasonable
to hold up the current PR until they are finished, as we hope it will recover
some of the increased compilation time experienced as a result of removing the
standardModuleSet "optimization"

TODO:
- check the docs for
  - [x] https://chapel-lang.org/docs/latest/builtins/ChapelIO.html
  - [x] https://chapel-lang.org/docs/latest/modules/standard/IO.html
  - [x] https://chapel-lang.org/docs/latest/primers/specialMethods.html
  - [x] https://chapel-lang.org/docs/latest/primers/learnChapelInYMinutes.html
- [x] fix for gasnet run as well
   - [x] also updated one file that ran with llvm (test/studies/dedup/dedup-externblock.chpl)
- [x] how to generate warning when user is relying on the IO module being included by default, so they have a release to update their code?
- [ ] run the twitter changes by the original author(s) to ensure they still capture the spirit of the tests.

### CODE REVIEW

**Comment by lydia-duncan** [comment]
This PR has just under 200 failures at the moment.  Making good progress!

**Comment by lydia-duncan** [comment]
clean gasnet run, clean standard paratest, but now I'm questioning whether everything triggers the error.  It seems like there are still tests that use `readln`, `openmem`, and other functions that are able to do so without a use of the IO module.  Vass pointed out a similar issue with the RangeChunk module (if all the uses were made private) - maybe this is related to #7200 ?

**Comment by lydia-duncan** [comment]
I am now convinced this is due to computeStandardModuleSet: it inserts the contents of used Standard modules into the root module, regardless of the use itself - so private uses and `use x only;` etc are not distinguished.

**Comment by lydia-duncan** [comment]
Current failures:
- need to insert a private use of the IO module when first class functions are defined, so that they can be printed
- update `modules/bradc/printModStuff/foo` and `modules/sungeun/init/printModuleInitOrder` again

File: modules/internal/ChapelLocale.chpl [hunk 1]
```diff
@@ -304,12 +304,6 @@ module ChapelLocale {
       return this;
     }
 
-    // A useful default definition is provided (not pure virtual).
-    pragma "no doc"
-    override proc writeThis(f) throws {
-      f <~> name;
-    }
-
```
Comment (mppf) [inline modules/internal/ChapelLocale.chpl]: Why are these moved to ChapelIO? It is OK for internal types but I'd just like to know user types / inheritance hierarchies can define these on their own.

File: modules/internal/LastResortIO.chpl [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+  compilerWarning("iohints will no longer be available by default, please\ninsert a use of the IO module to access it.");
+  return IO.iohints;
+}
+
+// TODO: file record methods
+pragma "last resort"
+proc file type {
+  compilerWarning("file will no longer be available by default, please\ninsert a use of the IO module to access it.");
+  return IO.file;
+}
```
Comment (mppf) [inline modules/internal/LastResortIO.chpl]: I was also expecting a `channel` type function here?

I feel pretty good about having wrappers only for:
 * types channel and file
 * non-method functions (free functions) - especially `open` and the like

File: modules/internal/LastResortIO.chpl [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+
+pragma "last resort"
+proc (IO.file).home ref: locale {
```
Comment (mppf) [inline modules/internal/LastResortIO.chpl]: Something about the ones like this one seems off/weird to me. E.g. if we resolved #14407 by saying that methods are available if a type is - wouldn't `myFile.home` already work?

This is part of the reason I think we should focus on making the wrapper-errors for types and free functions.

File: modules/standard/IO.chpl [hunk 4]
```diff
@@ -22,9 +22,6 @@
 
 Support for a variety of kinds of input and output.
 
-.. note:: All Chapel programs automatically ``use`` this module by default.
-          An explicit ``use`` statement is not necessary.
-
```
Comment (mppf) [inline modules/standard/IO.chpl]: I think instead this should list and link to the functions that are available by default

File: spec/Errors.tex [hunk 5]
```diff
@@ -130,6 +130,8 @@ \subsubsection{Catch clauses}
 \begin{chapelexample}{catching-errors.chpl}
 \begin{chapel}
 \begin{verbatim}
+use SysError;
```
Comment (mppf) [inline spec/Errors.tex]: Let's not modify the .tex spec since PR #14303 will remove it and it's not currently used

File: test/deprecated/io/checkopenmem.chpl [hunk 6]
```diff
@@ -0,0 +1,94 @@
+use IO except openmem; // to test deprecation warning for openmem
```
Comment (mppf) [inline test/deprecated/io/checkopenmem.chpl]: Shouldn't this test check the error messages for not putting `use IO` at all?

It seems unrealistic to have `use IO except openmem`.

Ditto for other tests in this directory.

File: modules/internal/ChapelLocale.chpl (hunk 1 above)
Comment (lydia-duncan) [reply modules/internal/ChapelLocale.chpl]: Having them declared here causes a circular dependency, since IO relies on locales.  Moving the writeThis function to ChapelIO allows us to wait to resolve those methods until the locales have been initialized (so it would be safe to initialize IO).

File: modules/internal/LastResortIO.chpl (hunk 2 above)
Comment (lydia-duncan) [reply modules/internal/LastResortIO.chpl]: The only reason I didn't was because I got tired and didn't want to hold up the PR any further.  I absolutely intend to get to doing that soon, but was hoping to merge this sooner so that we could see the compiler performance benefit (which I am now skeptical will happen, since I think the deprecation warnings are undoing the benefit not having the symbols caused, but that's another issue...)

Comment (mppf) [reply modules/internal/LastResortIO.chpl]: Right, but it seems to me to be sufficient to focus on the common types and free functions for the warning. That would address the problem of un-doing performance benefit, wouldn't it?

File: test/deprecated/io/checkopenmem.chpl (hunk 6 above)
Comment (lydia-duncan) [reply test/deprecated/io/checkopenmem.chpl]: The `open*` functions' arguments require types that trigger compilation errors instead of warnings, so I was trying to get their specific warning message.  Maybe that means I shouldn't bother adding compilation warnings for those functions at all, though, since in the basic case the user will need a use to call them in any real sense . . .

Comment (mppf) [reply test/deprecated/io/checkopenmem.chpl]: Right, I'd be in favor of trimming down the set of things we're adding warnings for (so that it covers almost all practical use cases)

**Comment by lydia-duncan** [comment]
> Do we need to make channel available by default, since it is used in writeThis/readWriteThis/etc?
Put another way - does one need to use IO in order to write a writeThis function? Are we happy with whatever outcome we end up with there?

I should also point out that the read/writeThis functions frequently use `<~>` and ioLiterals and those are also not available by default.  I think as a result we should require a use of IO to write a read/writeThis function

File: modules/standard/IO.chpl (hunk 4 above)
Comment (lydia-duncan) [reply modules/standard/IO.chpl]: How does the new note look?

Comment (mppf) [reply modules/standard/IO.chpl]: Maybe
```
.. note:: All Chapel programs automatically include :proc:`~ChapelIO.write`,
          :proc:`~ChapelIO.writeln` and :proc:`~ChapelIO.writef`. Please ``use IO`` to access other features in this module.
```

Comment (lydia-duncan) [reply modules/standard/IO.chpl]: We already automatically generate a sample use statement for standard modules.  IO was one of the few ones where we said "but actually, you don't need that use statement"

Comment (lydia-duncan) [reply modules/standard/IO.chpl]: It's not a burden to add it, but that's why I didn't do it on my own, so I'd like to check if that additional information is sufficient.  (This is how it looks on 1.20 today: https://chapel-lang.org/docs/latest/modules/standard/IO.html)

Comment (mppf) [reply modules/standard/IO.chpl]: OK, then what you have looks good to me.

File: modules/internal/LastResortIO.chpl (hunk 2 above)
Comment (lydia-duncan) [reply modules/internal/LastResortIO.chpl]: Looks like `channel` causes problems where `file` does not.  I think this is because ChapelIO needs channel for the functions and methods it contains and uses both LastResortIO and IO.  I can try making the use of LastResortIO slightly deeper (so inserting a fake module which uses it, and having ChapelIO use that fake module) but I would rather not place the use in someplace more prominent like ChapelStandard, for instance, since I think that would open a can of worms.

File: test/deprecated/io/checkopenmem.chpl (hunk 6 above)
Comment (lydia-duncan) [reply test/deprecated/io/checkopenmem.chpl]: Okay, I trimmed three tests and dropped the warning message for open.  I chose not to update the openmem test for two reasons:
1. That was going to be a lot of work and the point of me copying the test in the first place was to easily verify that the wrapped function behaved as expected rather than having a hidden error that a simple compilation would not detect
2. I had seen some problems where symbols that should have been allowed in when the use was limited in some way would not behave appropriately, so having a test when a limited use of IO was present would hopefully allow us to catch things a user would run into if they were already trying to limit what was brought it but had missed a symbol because we were including it by default.

Obviously reason 1 is mostly laziness, but I think reason 2 is sufficient to justify being lazy in this particular case, considering I updated the other ones.  Thoughts?

File: modules/internal/LastResortIO.chpl (hunk 2 above)
Comment (mppf) [reply modules/internal/LastResortIO.chpl]: Huh, I would think that the "last resort" flag would handle it. Maybe "last resort" doesn't work for paren-less functions for some reason?

File: test/deprecated/io/checkopenmem.chpl (hunk 6 above)
Comment (mppf) [reply test/deprecated/io/checkopenmem.chpl]: OK.

File: modules/internal/LastResortIO.chpl (hunk 2 above)
Comment (lydia-duncan) [reply modules/internal/LastResortIO.chpl]: We don't check against the "last resort" flag during scope resolution, which is when we're getting the error.  Perhaps we should?  Part of the point of that error message is to prevent the abuse I'm doing, it'd really be better for us to have a way to mark symbols other than functions as deprecated, as #14396 brings up
